docs/styleguide-cpp.md

```diff
@@ -1,320 +1,660 @@
+
+```c++
+class ExampleClass {
+public:
+    void first_public_method();
+    void second_public_method();
+    void third_public_method();
+
+protected:
+    void first_protected_method();
+    void second_protected_method();
+    void third_protected_method();
+
+private:
+    void first_private_method();
+    void second_private_method();
+    void third_private_method();
+};
+```
+
+#### 1.4.9. Methods which do not mutate an object’s state should be marked as a `const` method
+
+```c++
+//Correct
+bool RGBColour::is_colour_black() const {
+    return red_channel == 0 && green_channel == 0 && blue_channel == 0;
+}
+
+//Incorrect
+bool RGBColour::is_colour_black() {
+    return red_channel == 0 && green_channel == 0 && blue_channel == 0;
+}
+```
+
+#### 1.4.10. When argument or parameter lists are too long to easily fit on one line: begin the argument list on the next line, two indentation levels higher than the function’s indentation.
+
+```c++
+//Correct
+calling_a_function_with_a_very_long_name_and_many_arguments(
+        p_long_argument_name_one, p_long_argument_name_two, p_long_argument_name_three,
+        p_long_argument_name_four, p_long_argument_name_five, p_long_argument_name_six,
+        p_long_argument_name_seven, p_long_argument_name_eight);
+
+//Incorrect
+calling_a_function_with_a_very_long_name_and_many_arguments(p_long_argument_name_one, p_long_argument_name_two, p_long_argument_name_three, p_long_argument_name_four, p_long_argument_name_five, p_long_argument_name_six, p_long_argument_name_seven, p_long_argument_name_eight);
+```
+
+#### 1.4.11. In header files, put `#pragma once` at the top of the file, not C-style header guards
+
+```c++
+//Correct
+#pragma once
+/*
+header contents here
+*/
+
+//Incorrect
+#ifndef MY_COOL_HEADER_H
+#define MY_COOL_HEADER_H
+/*
+header contents here
+*/
+#endif
+```
+
+#### 1.4.12. Use enum classes, not plain enums
+
+```c++
+//Correct
+enum class Colours { RED, GREEN, BLUE, CHARCOAL_GREY };
+
+//Incorrect
+enum Colours { RED, GREEN, BLUE, CHARCOAL_GREY };
+```
+
+#### 1.4.13. Use type aliases, not \`typedef\`s
+
+```c++
+//Correct
+using MyContainer = std::vector<int>;
+using byte_t = unsigned char;
+
+//Incorrect
+typedef std::vector<int> MyContainer;
+typedef unsigned char byte_t;
+```
+
+#### 1.4.14. Use explicit types, do not use `auto`
+
+```c++
+//Correct
+godot::Vector2 coordinate = get_player_location();
+
+//Incorrect
+auto coordinate = get_player_location();
+```
+
+#### 1.4.15. Do not use C-style casts, instead use C++ casts (ie: `static_cast`, `const_cast`, `reinterpret_cast`)
+
+- `dynamic_cast` should only be used with caution
+
+```c++
+//Correct
+int my_number = static_cast<int>(some_float_value);
+
+//Incorrect
+int my_number = (int) some_float_value;
+```
+
+#### 1.4.16. When declaring a pointer/reference, put a space between the type and the asterisk/ambersand, put the asterisk/ambersand immediately before the variable name.
```

Comment:
Also subjective change

Reply:
I honestly don't much care for it either.

Reply:
It may be subjective, but my preference is still for `void *ptr`.